File: Data Structures & Algorithms/trapping-rain-water/submission-1.py
class Solution:
    def trap(self, heights: List[int]) -> int:
        l = 0
        r = len(heights) - 1
        max_l = heights[l]
        max_r = heights[r]
        res = 0
        while l < r:
            if max_l < max_r:
                l+=1
                max_l = max(max_l, heights[l])
                res+=max_l - heights[l]
            else:
                r-=1
                max_r = max(max_r, heights[r])
                res+=max_r - heights[r]
        
        return res

File: Data Structures & Algorithms/max-water-container/submission-1.cpp
#include <algorithm>
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0;
        int r = heights.size() - 1;
        int max_h = 0;
        while (l < r){
            max_h = max(max_h, (r - l) * min(heights[l], heights[r]));
            if (heights[l] < heights[r]){
                l++;
            }
            else{
                r--;
            }
        }
        return max_h;
    }
};
